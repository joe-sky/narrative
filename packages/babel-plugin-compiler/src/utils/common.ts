import type { NodePath } from '@babel/core';

export function displayError(path: NodePath, message: string) {
  let error: Error;
  try {
    error = path.buildCodeFrameError(message);
  } catch {
    error = new Error(message);
  }

  throw error;
}

export const NT_CONTROL_FLOW = '@narrative/control-flow';

export const IF = 'If';

export const SWITCH = 'Switch';

export const FOR = 'For';

export const SUB_TAGS_IF = {
  ELSE_IF: 'ElseIf',
  ELSE: 'Else'
};

export const ATTRS_IF = {
  WHEN: 'when'
};

export const SUB_TAGS_SWITCH = {
  CASE: 'Case',
  DEFAULT: 'Default'
};

export const ATTRS_SWITCH = {
  VALUE: 'value',
  IS: 'is',
  IN: 'in'
};

export const SUB_TAGS_FOR = {
  /**
   * ```tsx
   * <div>
   *   <For of={[1, 2, 3]}>
   *     {(item, { index, isFirst, isLast }) => <i key={index}>{item}</i>}
   *     <Empty>No Data</Empty>
   *   </For>
   * </div>
   *
   * Compiled ↓ ↓ ↓ ↓ ↓ ↓
   *
   * <div>
   *   {(__arr => {
   *     if (__arr?.length) {
   *       return __arr.map((item, index, arr) => {
   *         const length = arr.length;
   *         const isFirst = index === 0;
   *         const isLast = index === length - 1;
   *         return <i key={index}>{item}</i>;
   *       }, this);
   *     }
   *
   *     return 'No Data';
   *   })([1, 2, 3])}
   * </div>
   * ```
   */
  EMPTY: 'Empty'
};

export const ATTRS_FOR = {
  /**
   * ```tsx
   * <div>
   *   <For of={[1, 2, 3]}>
   *     {(item, { index, isFirst, isLast }) => <i key={index}>{item}</i>}
   *   </For>
   * </div>
   *
   * Compiled ↓ ↓ ↓ ↓ ↓ ↓
   *
   * <div>
   *   {[1, 2, 3]?.map((item, index, arr) => {
   *     const length = arr.length;
   *     const isFirst = index === 0;
   *     const isLast = index === length - 1;
   *     return <i key={index}>{item}</i>;
   *   }, this) || null}
   * </div>
   * ```
   */
  OF: 'of',

  /**
   * ```tsx
   * <div>
   *   <For in={{ a: 1, b: 2, c: 3 }}>
   *     {(item, { key }) => <i key={key}>{item}</i>}
   *     <Empty>No Data</Empty>
   *   </For>
   * </div>
   *
   * Compiled ↓ ↓ ↓ ↓ ↓ ↓
   *
   * <div>
   *   {(__obj => {
   *     const __keys = __obj ? Object.keys(__obj) : [];
   *
   *     if (__keys.length) {
   *       return __keys.map((key, index, arr) => {
   *         const item = __obj[key];
   *         const length = arr.length;
   *         const isFirst = index === 0;
   *         const isLast = index === length - 1;
   *         return <i key={key}>{item}</i>;
   *       }, this);
   *     }
   *
   *     return 'No Data';
   *   })({ a: 1, b: 2, c: 3 })}
   * </div>
   * ```
   */
  IN: 'in'
};

export const ARR_PARAM = '__arr';

export const OBJ_PARAM = '__obj';

export const KEYS_PARAM = '__keys';
