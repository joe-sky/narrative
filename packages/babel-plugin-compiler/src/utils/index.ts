import * as BabelCore from '@babel/core';

export const NT_CONTROL_FLOW = '@narrative/control-flow';

export type State = {
  get: (name: string) => any;
  set: (name: string, value: any) => any;
  opts: NtCompilerOptions;
  file: BabelCore.BabelFile;
};

export interface NtCompilerOptions {
  importedLib?: string[] | 'none';
}
