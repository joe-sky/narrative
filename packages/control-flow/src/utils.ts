export type JSXNode = JSX.Element | string | number | boolean | null | undefined | Record<any, any>;

export type JSXChild = JSXNode | JSXNode[];

export interface Childrenable {
  children?: JSXChild;
}
