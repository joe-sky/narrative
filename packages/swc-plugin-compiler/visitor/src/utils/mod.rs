pub mod ast;
pub mod common;