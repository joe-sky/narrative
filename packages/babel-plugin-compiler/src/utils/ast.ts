import * as types from '@babel/types';
import type {
  JSXElement,
  JSXIdentifier,
  JSXAttribute,
  JSXExpressionContainer,
  JSXText,
  JSXSpreadChild,
  JSXFragment,
  JSXEmptyExpression
} from '@babel/types';
import type { NodePath } from '@babel/traverse';
import { NT_CONTROL_FLOW } from './common';

export type JSXChild =
  | JSXText
  | JSXExpressionContainer
  | JSXSpreadChild
  | JSXElement
  | JSXFragment
  | JSXEmptyExpression
  | types.Expression;

export interface Attrs {
  condition: types.Expression;
  children: types.Expression;
}

export interface ForAttrs {
  source: types.Expression;
  callback: types.ArrowFunctionExpression;
  empty?: types.Expression;
}

export type FuncParam = types.Identifier | types.Pattern | types.RestElement;

export function isImportedByLib(
  identifier: string,
  path: NodePath<JSXElement>,
  libName: string | string[] = [NT_CONTROL_FLOW]
) {
  if (!Array.isArray(libName)) {
    libName = [libName];
    libName.push(NT_CONTROL_FLOW);
  }

  const bindingPath = path.scope?.getBinding(identifier)?.path;
  if (types.isImportDeclaration(bindingPath?.parent)) {
    const value = bindingPath?.parent.source.value;
    return value && libName.includes(value);
  }
}

export function getTagName(node: JSXElement) {
  return (node.openingElement.name as JSXIdentifier).name;
}

export function isTag(node: JSXElement, tagName: string) {
  return getTagName(node) === tagName;
}

export function getExpression(attribute: JSXAttribute) {
  if (types.isJSXExpressionContainer(attribute.value)) {
    return attribute.value.expression as types.Expression;
  } else {
    return attribute.value as types.Expression;
  }
}

export function getAttributeMap(node: JSXElement) {
  return node.openingElement.attributes.reduce((result, attr) => {
    if (types.isJSXAttribute(attr)) {
      result[((attr as JSXAttribute).name as JSXIdentifier).name] = attr;
    }
    return result;
  }, {} as Record<string, types.JSXAttribute>);
}

export function getKey(node: JSXElement): string | undefined {
  const key = getAttributeMap(node).key;
  return key ? (key.value as types.StringLiteral).value : undefined;
}

export function getChildren(node: JSXElement) {
  return types.react.buildChildren(node);
}

export function addKeyAttribute(node: JSXElement, keyValue: string | number) {
  let keyFound = false;

  node.openingElement.attributes.forEach(attrib => {
    if (types.isJSXAttribute(attrib) && attrib.name.name === 'key') {
      keyFound = true;
      return false;
    }
  });

  if (!keyFound) {
    const keyAttrib = types.jsxAttribute(types.jsxIdentifier('key'), types.stringLiteral('' + keyValue));
    node.openingElement.attributes.push(keyAttrib);
  }
}

export function convertChildrenToExpression(blocks: JSXChild[], keyPrefix?: string, callFunc?: boolean) {
  if (!blocks.length) {
    return types.nullLiteral();
  } else if (blocks.length === 1) {
    const firstBlock = blocks[0];

    if (keyPrefix && types.isJSXElement(firstBlock) && firstBlock.openingElement) {
      addKeyAttribute(firstBlock, keyPrefix);
    }

    if (callFunc && types.isArrowFunctionExpression(firstBlock)) {
      return types.callExpression(firstBlock, []);
    }

    return firstBlock as types.Expression;
  }

  for (let i = 0; i < blocks.length; i++) {
    const thisBlock = blocks[i];
    if (types.isJSXElement(thisBlock)) {
      const key = keyPrefix ? keyPrefix + '-' + i : i;
      addKeyAttribute(thisBlock, key);
    }
  }

  return types.arrayExpression(blocks as types.Expression[]);
}

export function findPropByName(obj: types.ObjectPattern, name: string) {
  return obj.properties.find(
    prop => types.isObjectProperty(prop) && types.isIdentifier(prop.key) && prop.key.name === name
  ) as types.ObjectProperty;
}

export function noopArrowFunctionExpression() {
  return types.arrowFunctionExpression([], types.blockStatement([]));
}
