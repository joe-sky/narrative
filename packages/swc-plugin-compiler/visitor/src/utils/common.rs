use swc_core::plugin::errors::HANDLER;
use swc_core::common::Span;
use tracing::error;

pub fn display_error(span: Span, message: &str) {
  error!(message);

  HANDLER.with(|handler| handler.span_err(span, message));
}

pub const NT_CONTROL_FLOW: &str = "@narrative/control-flow";

pub const IF: &str = "If";

pub const ELSE_IF: &str = "ElseIf";

pub const ELSE: &str = "Else";

pub const WHEN: &str = "when";

pub const SWITCH: &str = "Switch";

pub const CASE: &str = "Case";

pub const DEFAULT: &str = "Default";

pub const VALUE: &str = "value";

pub const IS: &str = "is";

pub const IN: &str = "in";

pub const FOR: &str = "For";

/// ```tsx
///
/// <div>
///   <For of={[1, 2, 3]}>
///     {(item, { index }) => <i key={index}>{item}</i>}
///     <Empty>No Data</Empty>
///   </For>
/// </div>
///
/// Compiled ↓ ↓ ↓ ↓ ↓ ↓
///
/// <div>
///   {(__arr => {
///     if (__arr?.length) {
///       return __arr.map((item, index, arr) => {
///         return <i key={index}>{item}</i>;
///       }, this);
///     }
///
///     return 'No Data';
///   })([1, 2, 3])}
/// </div>
/// ```
pub const EMPTY: &str = "Empty";

/// ```tsx
///
/// <div>
///   <For of={[1, 2, 3]}>
///     {(item, { index }) => <i key={index}>{item}</i>}
///   </For>
/// </div>
///
/// Compiled ↓ ↓ ↓ ↓ ↓ ↓
///
/// <div>
///   {[1, 2, 3]?.map?.((item, index, arr) => {
///     return <i key={index}>{item}</i>;
///   }, this) || null}
/// </div>
/// ```
pub const OF: &str = "of";

/// ```tsx
///
/// <div>
///   <For in={{ a: 1, b: 2, c: 3 }}>
///     {(item, { key }) => <i key={key}>{item}</i>}
///     <Empty>No Data</Empty>
///   </For>
/// </div>
///
/// Compiled ↓ ↓ ↓ ↓ ↓ ↓
///
/// <div>
///   {(__obj => {
///     const __keys = __obj ? Object.keys(__obj) : [];
///
///     if (__keys.length) {
///       return __keys.map(key => {
///         const item = __obj[key];
///         return <i key={key}>{item}</i>;
///       }, this);
///     }
///
///     return 'No Data';
///   })({ a: 1, b: 2, c: 3 })}
/// </div>
/// ```
pub const FOR_IN: &str = "in";

pub const ARR_PARAM: &str = "__arr";

pub const OBJ_PARAM: &str = "__obj";

pub const KEYS_PARAM: &str = "__keys";