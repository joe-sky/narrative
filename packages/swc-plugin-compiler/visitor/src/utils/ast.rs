use swc_core::common::{ DUMMY_SP, Spanned };
use swc_core::ecma::ast::{
  Expr,
  Ident,
  JSXAttr,
  JSXAttrName,
  JSXAttrOrSpread,
  JSXAttrValue,
  JSXClosingFragment,
  JSXElement,
  JSXElementChild,
  JSXElementName,
  JSXExpr,
  JSXExprContainer,
  JSXFragment,
  JSXOpeningFragment,
  JSXText,
  Lit,
  Null,
  Str,
  BinExpr,
  op,
  CallExpr,
  Callee,
  MemberExpr,
  MemberProp,
  ExprOrSpread,
};
use swc_core::ecma::atoms::JsWord;

use super::common::{ display_error, WHEN, IS, IN, VALUE, OF, FOR_IN };

pub fn clone_children(children: &Vec<JSXElementChild>) -> Vec<JSXElementChild> {
  let mut copy: Vec<JSXElementChild> = Vec::new();

  for child in children {
    match child {
      JSXElementChild::JSXText(JSXText { value, .. }) => {
        let mut value = value.to_string();

        value = value.replace('\n', "");

        if value.trim() != "" {
          copy.push((*child).clone());
        }
      }
      _ => {
        copy.push((*child).clone());
      }
    }
  }

  copy
}

pub fn convert_children_to_expression(children: Vec<JSXElementChild>) -> Expr {
  match children.len() {
    0 => Expr::Lit(Lit::Null(Null { span: DUMMY_SP })),
    1 => {
      let sub_element = &children[0];

      match sub_element {
        JSXElementChild::JSXElement(value) => Expr::JSXElement(Box::new((**value).clone())),
        JSXElementChild::JSXExprContainer(JSXExprContainer { expr: JSXExpr::Expr(expr), .. }) => {
          match &**expr {
            Expr::Arrow(array_expr) =>
              Expr::Call(CallExpr {
                callee: Callee::Expr(Box::new(Expr::Arrow((*array_expr).clone()))),
                args: vec![],
                type_args: None,
                span: DUMMY_SP,
              }),
            _ => (**expr).clone(),
          }
        }
        JSXElementChild::JSXText(JSXText { value, .. }) => {
          let mut content = value.to_string();

          content = content.replace('\n', "");

          Expr::Lit(
            Lit::Str(Str {
              span: DUMMY_SP,
              value: content.trim().into(),
              raw: None,
            })
          )
        }
        _ => Expr::Lit(Lit::Null(Null { span: DUMMY_SP })),
      }
    }
    _ =>
      Expr::JSXFragment(JSXFragment {
        span: DUMMY_SP,
        children,
        opening: JSXOpeningFragment { span: DUMMY_SP },
        closing: JSXClosingFragment { span: DUMMY_SP },
      }),
  }
}

pub fn get_when_expression(jsx_element: &JSXElement) -> Expr {
  jsx_element.opening.attrs
    .iter()
    .find(|attr| {
      if let JSXAttrOrSpread::JSXAttr(JSXAttr { name: JSXAttrName::Ident(Ident { sym, span, .. }), .. }) = attr {
        if sym == WHEN {
          return true;
        } else {
          display_error(*span, format!("Only \"when\" attribute allowed, got: \"{}\".", sym).as_str());
        }
      }

      false
    })
    .map(|attr| {
      match attr {
        JSXAttrOrSpread::JSXAttr(JSXAttr { value, .. }) =>
          match value {
            Some(JSXAttrValue::JSXExprContainer(value)) => {
              let JSXExprContainer { expr, .. } = value;

              match expr {
                JSXExpr::Expr(value) => (**value).clone(),
                _ => Expr::Lit(Lit::Bool(false.into())),
              }
            }
            _ => Expr::Lit(Lit::Bool(false.into())),
          }
        JSXAttrOrSpread::SpreadElement(value) => {
          display_error(value.dot3_token.span(), "Spread is invalid for the value of when.");

          Expr::Lit(Lit::Bool(false.into()))
        }
      }
    })
    .unwrap_or_else(|| {
      let element_name = get_jsx_element_name(&jsx_element.opening.name);

      display_error(
        jsx_element.opening.span,
        format!("Attribute \"when\" is required for the <{}> tag.", element_name).as_str()
      );

      Expr::Lit(Lit::Bool(false.into()))
    })
}

pub fn get_is_expression(jsx_element: &JSXElement, parent_jsx_element: &JSXElement) -> Expr {
  let mut has_in = false;

  jsx_element.opening.attrs
    .iter()
    .find(|attr| {
      if let JSXAttrOrSpread::JSXAttr(JSXAttr { name: JSXAttrName::Ident(Ident { sym, span, .. }), .. }) = attr {
        if sym == IS || sym == IN {
          if sym == IN {
            has_in = true;
          }

          return true;
        } else {
          display_error(*span, format!("Only \"is\" or \"in\" attribute allowed, got: \"{}\".", sym).as_str());
        }
      }

      false
    })
    .map(|attr| {
      match attr {
        JSXAttrOrSpread::JSXAttr(JSXAttr { value, .. }) =>
          match value {
            Some(JSXAttrValue::JSXExprContainer(value)) => {
              let JSXExprContainer { expr, .. } = value;

              match expr {
                JSXExpr::Expr(expr_value) => {
                  parent_jsx_element.opening.attrs
                    .iter()
                    .find(|attr| {
                      if
                        let JSXAttrOrSpread::JSXAttr(
                          JSXAttr { name: JSXAttrName::Ident(Ident { sym, span, .. }), .. },
                        ) = attr
                      {
                        if sym == VALUE {
                          return true;
                        } else {
                          display_error(*span, format!("Only \"value\" attribute allowed, got: \"{}\".", sym).as_str());
                        }
                      }

                      false
                    })
                    .map(|attr| {
                      match attr {
                        JSXAttrOrSpread::JSXAttr(JSXAttr { value, .. }) =>
                          match value {
                            Some(JSXAttrValue::JSXExprContainer(value)) => {
                              let JSXExprContainer { expr, .. } = value;

                              match expr {
                                JSXExpr::Expr(parent_expr_value) => {
                                  if has_in {
                                    Expr::Call(CallExpr {
                                      callee: Callee::Expr(
                                        Box::new(
                                          Expr::Member(MemberExpr {
                                            obj: (*expr_value).clone(),
                                            prop: MemberProp::Ident(Ident {
                                              sym: JsWord::from("includes"),
                                              optional: false,
                                              span: DUMMY_SP,
                                            }),
                                            span: DUMMY_SP,
                                          })
                                        )
                                      ),
                                      args: vec![ExprOrSpread { expr: (*parent_expr_value).clone(), spread: None }],
                                      type_args: None,
                                      span: DUMMY_SP,
                                    })
                                  } else {
                                    Expr::Bin(BinExpr {
                                      op: op!("==="),
                                      left: (*parent_expr_value).clone(),
                                      right: (*expr_value).clone(),
                                      span: DUMMY_SP,
                                    })
                                  }
                                }
                                _ => Expr::Lit(Lit::Bool(false.into())),
                              }
                            }
                            _ => Expr::Lit(Lit::Bool(false.into())),
                          }
                        JSXAttrOrSpread::SpreadElement(value) => {
                          display_error(value.dot3_token.span(), "Spread is invalid for the value of \"value\".");

                          Expr::Lit(Lit::Bool(false.into()))
                        }
                      }
                    })
                    .unwrap_or_else(|| {
                      let element_name = get_jsx_element_name(&parent_jsx_element.opening.name);

                      display_error(
                        parent_jsx_element.opening.span,
                        format!("Attribute \"value\" is required for the <{}> tag.", element_name).as_str()
                      );

                      Expr::Lit(Lit::Bool(false.into()))
                    })
                }
                _ => Expr::Lit(Lit::Bool(false.into())),
              }
            }
            _ => Expr::Lit(Lit::Bool(false.into())),
          }
        JSXAttrOrSpread::SpreadElement(value) => {
          display_error(
            value.dot3_token.span(),
            format!("Spread is invalid for the value of \"{}\".", if has_in { IN } else { IS }).as_str()
          );

          Expr::Lit(Lit::Bool(false.into()))
        }
      }
    })
    .unwrap_or_else(|| {
      let element_name = get_jsx_element_name(&jsx_element.opening.name);

      display_error(
        jsx_element.opening.span,
        format!("Attribute \"is\" or \"in\" is required for the <{}> tag.", element_name).as_str()
      );

      Expr::Lit(Lit::Bool(false.into()))
    })
}

pub fn get_of_expression(jsx_element: &JSXElement) -> (Expr, bool) {
  let mut has_in = false;

  let source = jsx_element.opening.attrs
    .iter()
    .find(|attr| {
      if let JSXAttrOrSpread::JSXAttr(JSXAttr { name: JSXAttrName::Ident(Ident { sym, span, .. }), .. }) = attr {
        if sym == OF || sym == FOR_IN {
          if sym == FOR_IN {
            has_in = true;
          }

          return true;
        } else {
          display_error(*span, format!("Only \"of\" or \"in\" attribute allowed, got: \"{}\".", sym).as_str());
        }
      }

      false
    })
    .map(|attr| {
      match attr {
        JSXAttrOrSpread::JSXAttr(JSXAttr { value, .. }) =>
          match value {
            Some(JSXAttrValue::JSXExprContainer(value)) => {
              let JSXExprContainer { expr, .. } = value;

              match expr {
                JSXExpr::Expr(value) => (**value).clone(),
                _ => Expr::Lit(Lit::Bool(false.into())),
              }
            }
            _ => Expr::Lit(Lit::Bool(false.into())),
          }
        JSXAttrOrSpread::SpreadElement(value) => {
          display_error(
            value.dot3_token.span(),
            format!("Spread is invalid for the value of \"{}\".", if has_in { FOR_IN } else { OF }).as_str()
          );

          Expr::Lit(Lit::Bool(false.into()))
        }
      }
    })
    .unwrap_or_else(|| {
      let element_name = get_jsx_element_name(&jsx_element.opening.name);

      display_error(
        jsx_element.opening.span,
        format!("Attribute \"of\" or \"in\" is required for the <{}> tag.", element_name).as_str()
      );

      Expr::Lit(Lit::Bool(false.into()))
    });

  (source, has_in)
}

pub fn get_jsx_element_name(jsx_element_name: &JSXElementName) -> &str {
  match jsx_element_name {
    JSXElementName::Ident(Ident { sym, .. }) => sym,
    _ => "unknown",
  }
}

pub fn wrap_by_child_jsx_expr_container(expr: Expr) -> JSXElementChild {
  JSXElementChild::JSXExprContainer(JSXExprContainer {
    span: DUMMY_SP,
    expr: JSXExpr::Expr(Box::new(expr)),
  })
}

pub fn null_literal() -> Expr {
  Expr::Lit(Lit::Null(Null { span: DUMMY_SP }))
}