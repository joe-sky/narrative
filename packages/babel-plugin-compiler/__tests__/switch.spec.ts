import { Test, transpile } from './utils';

const switchTests: Test[] = [
  {
    name: 'simple switch',
    from: `
      <Switch value={props.value}>
        <Case is={1}>test1</Case>
        <Case is={2}>test2</Case>
        <Case is={3}>test3</Case>
      </Switch>
    `
  },
  {
    name: 'nesting switch',
    from: `
    <Switch value={props.value}>
      <Case is={1}>
        <Switch value={props.value}>
          <Case is={1}>test1</Case>
          <Case is={2}>test2</Case>
          <Case is={3}>test3</Case>
        </Switch>
      </Case>
      <Case is={2}>
        <Switch value={props.value}>
          <Case is={1}>test1</Case>
          <Case is={2}>test2</Case>
          <Case is={3}>test3</Case>
        </Switch>
      </Case>
      <Case is={3}>
        <Switch value={props.value}>
          <Case is={1}>test1</Case>
          <Case is={2}>test2</Case>
          <Case is={3}>test3</Case>
        </Switch>
      </Case>
    </Switch>
    `
  },
  {
    name: 'has values',
    from: `
      <Switch value={props.value}>
        <Case is={1}>test1</Case>
        <Case is={2}>test2</Case>
        <Case in={[3, 4, 5]}>test3</Case>
      </Switch>
    `
  },
  {
    name: 'has default',
    from: `
      <Switch value={props.value}>
        <Case is={1}>test1</Case>
        <Case is={2}>test2</Case>
        <Case is={3}>test3</Case>
        <Default>test</Default>
      </Switch>
    `
  },
  {
    name: 'function children',
    from: `
      <Switch value={props.value}>
        <Case is={1}>{() => 'test1'}</Case>
        <Case is={2}>{() => 'test2'}</Case>
        <Case is={3}>{() => 'test3'}</Case>
        <Default>{() => 'test'}</Default>
      </Switch>
    `
  },
  {
    name: 'if wrapped in Fragments',
    from: `
      const jsx = <>
        <div></div>
        <Switch value={props.value}>
          <Case is={1}>test1</Case>
          <Case is={2}>test2</Case>
          <Case is={3}>test3</Case>
        </Switch>
        <div></div>
      </>
    `
  }
];

switchTests.forEach(({ name, from }) => {
  test(name, async () => {
    expect(await transpile(from)).toMatchSnapshot(name);
  });
});
