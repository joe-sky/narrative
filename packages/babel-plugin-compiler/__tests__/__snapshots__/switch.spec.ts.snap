// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function children: function children 1`] = `"1 === props.value ? (() => 'test1')() : 2 === props.value ? (() => 'test2')() : 3 === props.value ? (() => 'test3')() : (() => 'test')();"`;

exports[`has default: has default 1`] = `"1 === props.value ? \\"test1\\" : 2 === props.value ? \\"test2\\" : 3 === props.value ? \\"test3\\" : \\"test\\";"`;

exports[`has values: has values 1`] = `"1 === props.value ? \\"test1\\" : 2 === props.value ? \\"test2\\" : [3, 4, 5].includes(props.value) ? \\"test3\\" : null;"`;

exports[`if wrapped in Fragments: if wrapped in Fragments 1`] = `
"const jsx = <>
        <div></div>
        {1 === props.value ? \\"test1\\" : 2 === props.value ? \\"test2\\" : 3 === props.value ? \\"test3\\" : null}
        <div></div>
      </>;"
`;

exports[`nesting switch: nesting switch 1`] = `"1 === props.value ? 1 === props.value ? \\"test1\\" : 2 === props.value ? \\"test2\\" : 3 === props.value ? \\"test3\\" : null : 2 === props.value ? 1 === props.value ? \\"test1\\" : 2 === props.value ? \\"test2\\" : 3 === props.value ? \\"test3\\" : null : 3 === props.value ? 1 === props.value ? \\"test1\\" : 2 === props.value ? \\"test2\\" : 3 === props.value ? \\"test3\\" : null : null;"`;

exports[`simple switch: simple switch 1`] = `"1 === props.value ? \\"test1\\" : 2 === props.value ? \\"test2\\" : 3 === props.value ? \\"test3\\" : null;"`;
